qiskit/circuit/library/grover_operator.py

```diff
@@ -0,0 +1,187 @@
+    and an input state $\mathcal{A}$. For the textbook Grover search, $\mathcal{A} = H^{\otimes n}$,
+    however for generic amplitude amplification the input state might differ [3].
+    With these terms, the Grover operator can be written as
+
+    .. math::
+
+        \mathcal{Q} = \mathcal{A} \mathcal{S}_0 \mathcal{A} \mathcal{S_f}
+
+    .. note::
+
+        Sometimes the Grover operator is defined with a negative sign, in that case
+        $\mathcal{S}_0$ multiplies all states *except* $\ket{0}^{\otimes n}$ with -1.
+        In our formulation, $\mathcal{S}_0$ only multiplies $\ket{0}^{\otimes n}$ with -1.
+
+    Examples:
+        >>> from qiskit.circuit import QuantumCircuit
+        >>> from qiskit.circuit.library import GroverOperator
+        >>> oracle = QuantumCircuit(2)
+        >>> oracle.z(0)  # good state = first qubit is |1>
+        >>> grover_op = GroverOperator(oracle, insert_barriers=True)
+        >>> grover_op.draw()
+                 ┌───┐ ░ ┌───┐ ░ ┌───┐          ┌───┐      ░ ┌───┐
+        state_0: ┤ Z ├─░─┤ H ├─░─┤ X ├───────■──┤ X ├──────░─┤ H ├
+                 └───┘ ░ ├───┤ ░ ├───┤┌───┐┌─┴─┐├───┤┌───┐ ░ ├───┤
+        state_1: ──────░─┤ H ├─░─┤ X ├┤ H ├┤ X ├┤ H ├┤ X ├─░─┤ H ├
+                       ░ └───┘ ░ └───┘└───┘└───┘└───┘└───┘ ░ └───┘
+
+        >>> oracle = QuantumCircuit(1)
+        >>> oracle.z(0)  # the qubit state |1> is the good state
+        >>> state_in = QuantumCircuit(1)
+        >>> state_in.ry(0.2, 0)  # non-uniform state preparation
+        >>> grover_op = GroverOperator(oracle, state_in)
+        >>> grover_op.draw()
+                 ┌───┐┌──────────┐┌───┐┌───┐┌───┐┌─────────┐
+        state_0: ┤ Z ├┤ RY(-0.2) ├┤ X ├┤ Z ├┤ X ├┤ RY(0.2) ├
+                 └───┘└──────────┘└───┘└───┘└───┘└─────────┘
+
+    References:
+        [1]: L. K. Grover (1996), A fast quantum mechanical algorithm for database search,
+            `arXiv:quant-ph/9605043 <https://arxiv.org/abs/quant-ph/9605043>`_.
+        [2]: I. Chuang & M. Nielsen, Quantum Computation and Quantum Information,
+            Cambridge: Cambridge University Press, 2000. Chapter 6.1.2.
+        [3]: Brassard, G., Hoyer, P., Mosca, M., & Tapp, A. (2000).
+            Quantum Amplitude Amplification and Estimation.
+            `arXiv:quant-ph/0005055 <http://arxiv.org/abs/quant-ph/0005055>`_.
+    """
+
+    def __init__(self, oracle: QuantumCircuit,
+                 state_in: Optional[QuantumCircuit] = None,
+                 zero_reflection: Optional[QuantumCircuit] = None,
+                 idle_qubits: Optional[List[int]] = None,
+                 insert_barriers: bool = False,
+                 mcx: str = 'noancilla',
+                 name: str = 'Q') -> None:
+        """
+        Args:
+            oracle: The oracle implementing a reflection about the bad state.
+            state_in: The operator preparing the good and bad state. For Grover's algorithm,
```

Comment:
why is the `state_in` part of the "grover operator"? shouldn't the operator be independent from what you feed it?

Reply:
In the standard formulation of Grover's algorithm you use an equal superposition/H-layer as input state, but you in the more general formulation of amplitude amplification you can provide a different one. It's needed in the operator itself because you need to apply the inverse of it again. 
Grover: `H Mark H Oracle`
Amplitude Amplification: `A_dg Mark A Oracle` (where `A = state_in`)

Maybe `state_in` is not an ideal name? 🤔

Reply:
I can think of some other potential names such as `state_preparation` / `starting_superposition` / `list_initializer` but none that I think are objectively clearer than `state_in`. I think the docstring is clear for people that require this, and those that don't can safely ignore it.

Reply:
`state_preparation` seems like a good name 👍